### PR TITLE
Heuristic for `segment_matmul`

### PR DESCRIPTION
**Pull request by puririshi98**
good to go

### CODE REVIEW

**Comment by puririshi98** [comment]
testing the hueristic for linear shows the hueristic is highly useful:
test: https://github.com/puririshi98/rgcn_pyg_lib_forward_bench/blob/main/heterolinear_bench.py
results:
```
Loop Times: [0.0010712862014770508, 0.001991710662841797, 0.0033931493759155273, 0.006711206436157227, 0.013364143371582031, 0.026939492225646972, 0.05452749252319336, 0.1401316976547241, 0.4013452672958374]
Fused Times: [0.0008774948120117188, 0.001725149154663086, 0.0030545616149902344, 0.005789413452148438, 0.011584835052490234, 0.02313891887664795, 0.0058453369140625, 0.010598626136779785, 0.0210475492477417]
```

File: torch_geometric/utils/hetero.py [hunk 1]
```diff
@@ -9,6 +9,24 @@
 from torch_geometric.utils.num_nodes import maybe_num_nodes_dict
 
 
+def segmatmul_hueristic(inputs, ptr, others):
```
Comment (rusty1s) [inline torch_geometric/utils/hetero.py:12]: Can we also push the code that was used to learn this?

File: torch_geometric/nn/to_hetero_module.py [hunk 2]
```diff
@@ -54,7 +55,7 @@ def dict_forward(
 
         if not torch_geometric.typing.WITH_PYG_LIB:
             return {
-                key: self.hetero_module.lins[i](x_dict[key])
```
Comment (rusty1s) [inline torch_geometric/nn/to_hetero_module.py:57]: Why change? We lose the `bias` here.

File: torch_geometric/nn/dense/linear.py [hunk 3]
```diff
@@ -275,11 +270,10 @@ def forward(self, x: Tensor, type_vec: Tensor) -> Tensor:
                 out_unsorted[perm] = out
                 out = out_unsorted
         else:
-            assert self.lins is not None
             out = x.new_empty(x.size(0), self.out_channels)
-            for i, lin in enumerate(self.lins):
+            for i in range(self.num_types):
                 mask = type_vec == i
-                out[mask] = lin(x[mask])
+                out[mask] = F.linear(x[mask], self.weight[i].T)
```
Comment (rusty1s) [inline torch_geometric/nn/dense/linear.py:276]: `bias`?

File: torch_geometric/utils/hetero.py [hunk 4]
```diff
@@ -9,6 +9,109 @@
...
+    #     [  125.11603189 12133.21523472   163.81222321    32.43755536]
+    # scaler scale=
+    #     [  163.34480422 27572.94543809   177.6426489     56.82103934]
+    # svm weights=
+    #     [[ 2.43877659e+00  1.67583047e+00 -5.20527282e-04  3.43925501e-01]]
+    # svm bias=
+    #     [1.20236999]
+
+
+def segmatmul_hueristic(inputs, ptr, others):
```
Comment (Kh4L) [inline torch_geometric/utils/hetero.py:97]:
```suggestion
def segmatmul_heuristic(inputs, ptr, others):
```

File: torch_geometric/nn/dense/linear.py [hunk 5]
```diff
@@ -275,11 +270,12 @@ def forward(self, x: Tensor, type_vec: Tensor) -> Tensor:
                 out_unsorted[perm] = out
                 out = out_unsorted
         else:
-            assert self.lins is not None
             out = x.new_empty(x.size(0), self.out_channels)
-            for i, lin in enumerate(self.lins):
+            for i in range(self.num_types):
```
Comment (wsad1) [inline torch_geometric/nn/dense/linear.py:274]: What happens here if a given `node_type` isn't present in the batch?

File: torch_geometric/nn/dense/linear.py [hunk 6]
```diff
@@ -215,48 +216,39 @@ def __init__(self, in_channels: int, out_channels: int, num_types: int,
...
     def forward(self, x: Tensor, type_vec: Tensor) -> Tensor:
         r"""
         Args:
             x (torch.Tensor): The input features.
             type_vec (torch.Tensor): A vector that maps each entry to a type.
         """
-        if torch_geometric.typing.WITH_PYG_LIB:
+        if x.size(0) == 1:
+            # for single vector inference don't use segmm
+            self.use_segmm = 0
```
Comment (wsad1) [inline torch_geometric/nn/dense/linear.py:248]: Do you want to reset this to its original value at the end. If not if the user passes larger values later, `use_segmm` will be stuck at 0.
Also what if we just don't have this logic, does it cause bugs? 
```suggestion
```

File: torch_geometric/nn/conv/rgcn_conv.py [hunk 7]
```diff
@@ -230,14 +230,18 @@ def forward(self, x: Union[OptTensor, Tuple[OptTensor, Tensor]],
...
+                                                  or bool(self.use_segmm)):
                 if not self.is_sorted:
                     if (edge_type[1:] < edge_type[:-1]).any():
                         edge_type, perm = index_sort(
                             edge_type, max_value=self.num_relations)
                         edge_index = edge_index[:, perm]
                 edge_type_ptr = index2ptr(edge_type, self.num_relations)
+                if self.use_segmm == -1:
+                    self.use_segmm = segmatmul_heuristic(
+                        x_l, edge_type, self.weight)
```
Comment (wsad1) [inline torch_geometric/nn/conv/rgcn_conv.py:244]:
```suggestion
                        x_l, edge_type_ptr, self.weight)
```

File: torch_geometric/utils/hetero.py [hunk 8]
```diff
@@ -9,6 +9,109 @@
...
+    #     [  125.11603189 12133.21523472   163.81222321    32.43755536]
+    # scaler scale=
+    #     [  163.34480422 27572.94543809   177.6426489     56.82103934]
+    # svm weights=
+    #     [[ 2.43877659e+00  1.67583047e+00 -5.20527282e-04  3.43925501e-01]]
+    # svm bias=
+    #     [1.20236999]
+
+
+def segmatmul_heuristic(inputs, ptr, others):
```
Comment (wsad1) [inline torch_geometric/utils/hetero.py:97]:
```suggestion
def segmatmul_heuristic(x: Tensor, type_ptr, weight: Tensor):
```

**Review by wsad1 — COMMENTED** [review]
Nice work.

File: torch_geometric/nn/dense/linear.py (hunk 6 above)
Comment (puririshi98) [reply torch_geometric/nn/dense/linear.py:248]: i think we can leave this off tbh, the hueristic should choose the right choice

**Review by wsad1 — APPROVED** [review]
Thanks for the good work.

**Pull request merged**